pkg/gadgets/profile/tcprtt/tracer/tracer.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -type hist -cc clang tcpRTT ./bpf/tcprtt.bpf.c -- -I./bpf/ -I../../../../${TARGET}
+
+type Config struct {
+	milliseconds        bool
+	extension           bool
+	localAddrHist       bool
+	remoteAddrHist      bool
+	filterLocalAddress  uint32
+	filterRemoteAddress uint32
+}
+
+type Tracer struct {
+	objs                tcpRTTObjects
+	tcpRcvEstKprobeLink link.Link
+	tcpRcvEstFentryLink link.Link
+
+	config *Config
+	logger logger.Logger
+}
+
+func (t *Tracer) RunWithResult(gadgetCtx gadgets.GadgetContext) ([]byte, error) {
+	t.logger = gadgetCtx.Logger()
+
+	if err := t.parseParams(gadgetCtx.GadgetParams()); err != nil {
+		return nil, fmt.Errorf("parsing parameters: %w", err)
+	}
+
+	defer t.close()
+	if err := t.install(); err != nil {
+		return nil, fmt.Errorf("installing tracer: %w", err)
+	}
+
+	gadgetcontext.WaitForTimeoutOrDone(gadgetCtx)
+
+	result, err := t.collectResult()
+	if err != nil {
+		return nil, fmt.Errorf("collecting result: %w", err)
+	}
+	return result, nil
+}
+
+// ipStringToUint32 converts an IP address (IPv4 only) string to a uint32 in
+// little endian format.
+func ipStringToUint32(ipAddr string) (uint32, error) {
+	ip := net.ParseIP(ipAddr)
+	if ip == nil {
+		return 0, fmt.Errorf("invalid IP address: %s", ipAddr)
+	}
+	ip = ip.To4()
+	return binary.LittleEndian.Uint32(ip), nil
+}
+
+func (t *Tracer) parseParams(params *params.Params) error {
+	t.config.milliseconds = params.Get(ParamMilliseconds).AsBool()
+	t.config.extension = params.Get(ParamExtension).AsBool()
+
+	t.config.localAddrHist = params.Get(ParamByLocalAddress).AsBool()
+	t.config.remoteAddrHist = params.Get(ParamByRemoteAddress).AsBool()
+	if t.config.localAddrHist && t.config.remoteAddrHist {
+		return fmt.Errorf("local and remote address histograms cannot be enabled at the same time")
+	}
+
+	lAddr := params.Get(ParamFilterLocalAddress).AsString()
+	if lAddr != "" {
+		l, err := ipStringToUint32(lAddr)
+		if err != nil {
+			return fmt.Errorf("parsing local address: %w", err)
+		}
+		t.config.filterLocalAddress = l
+		t.logger.Debugf("Filtering for local address %s (%d)", lAddr, t.config.filterLocalAddress)
+	}
+
+	rAddr := params.Get(ParamFilterRemoteAddress).AsString()
+	if rAddr != "" {
+		r, err := ipStringToUint32(rAddr)
+		if err != nil {
+			return fmt.Errorf("parsing remote address: %w", err)
+		}
+		t.config.filterRemoteAddress = r
+		t.logger.Debugf("Filtering for remote address %s (%d)", rAddr, t.config.filterRemoteAddress)
+	}
+
+	if (t.config.filterLocalAddress != 0 || t.config.filterRemoteAddress != 0) &&
+		(t.config.localAddrHist || t.config.remoteAddrHist) {
+		return fmt.Errorf("filtering by local/remote address cannot be enabled at the same time of showing histograms by local/remote address")
+	}
```

Comment:
What's the reason for this limitation? I think there are some valid use cases:
- Show histogram by local address to all connections going to 1.1.1.1: `ig profile tcprtt --byladdr --raddr 1.1.1.1`
- `sudo ./ig profile tcprtt --byraddr --raddr 1.1.1.1`: I think it's implicit but I don't see why we should print an error, it should "just work".

Reply:
You're right. I can't remember why I limited it. I added it back and added some tests for it.